chore!: Add deprecation notices to values in blockly.js

maribethb opened the pull request:
<!--
  - Thanks for submitting code to Blockly!  Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## The basics

<!-- TODO: Verify the following, checking each box with an 'x' between the brackets: [x] -->

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

## The details
### Resolves

### Proposed Changes

- Adds deprecation notices to aliased functions and some properties (see below) in blockly.js
- Moves the few functions that aren't aliased to the top of the file for readability
- Uses a consistent style for deprecated functions. Added the `@deprecated` tag, `@see` tag to link the documentation, and the warning.
- Moves the aliased properties to use accessors so that the warning can be added.
- Updates all internal usages of the properties changed in this PR, including in tests and demos.
- Makes clipboard functions public, as discussed offline.


### Test Coverage

Tested with `npm run test`, in the playground, and ran the demos (with updated built files) to make sure there are no console warnings about deprecated functions.

### Documentation

None

### Additional Information

Did not move all of the aliases at the bottom of the file because I suspect some of them should not actually be public and some of them will need setters. See #5554 

Make sure you expand the changes in blockly.js as GitHub is hiding them by default.

BeksOmega commented:
@maribethb I don't think I'll have time to review this today :/ could you assign this to someone else?

alschmiedt commented on core/blockly.js:
```diff
@@ -261,13 +330,110 @@ Object.defineProperties(exports, {
       common.setSelected(newSelection);
     }
   },
+  ALIGN_LEFT: {
```
In #5567 @cpcallen has started adding a jsdoc above these so that they will have the appropriate docs.

alschmiedt commented on tests/mocha/procedures_test_helpers.js:
```diff
@@ -5,6 +5,8 @@
 goog.module('Blockly.test.procedureHelpers');
 
+const {ConnectionType} = goog.require('Blockly.ConnectionType');
+
```
Add one more new line.

maribethb commented:
Update: Merged in develop, fixed updated references (Input alignments moved in the meantime), updated all September dates in blockly.js to be December instead, added jsdoc to the properties, etc.

The only thing I'm not sure about is that previously we had agreed `Blockly.defineBlocksWithJsonArray` and `Blockly.isNumber` would not be deprecated, but now they have both moved out of this file. However `isNumber` was marked deprecated but `defineBlocksWithJsonArray` was not. Not sure what our intentions here, so I did not add the deprecation warning to `defineBlocksWithJsonArray` but I can if we want to.

cpcallen commented:
As I understood it, most of the code that was moved from `Blockly` to `Blockly.common` (and other places) was done to resolve circular `requires` and (more recently) so that no `goog.module` need depend directly on the `Blockly` module, rather than because we wanted to change the Blockly API as such—i.e., this code movement was largely an implementation detail.

Of course there will most probably be intentional API changes in the (hopefully near) future as we move towards TypeScript, and I am strongly in favour of trying to make the Blockly API as cleanly organised and modular as possible, but I think it would be well worth nailing down (at least in broad strokes) what we will want that future Blockly API to look like before we start forcing external developers to make arbitrary changes to their code by deprecating everything on `Blockly`.

E.g.: moving the implementation of `defineBlocksWithJsonArray` to `Blockly.common` let us fix problems that would otherwise have occurred when migrating the `blocks/*.js` files, but I didn't understand this to mean that we intended to promulgate a change to the API: there's no obvious reason why external code should not continue to use `Blockly.defineBlocksWithJsonArray` at the moment.

This actually holds true for almost all the other items on `Blockly` deprecated in this and other recent PRs, with the exception of property accessors, where (AFAIK) we do want external code to switch to using the new get/set functions instead—though even here there is no reason in principle that the new functions could not be re-exported from `blockly.js` if we wanted them to be, rather than being called via various sub-modules.

While I am generally in favour of encouraging external developers to update their code as early as possible, I think there is not enough clarity about what the future Blockly API will look like to justify asking developers to make (any but a few) changes at the present time.

cpcallen reviewed (changes requested):
As I understood it, most of the code that was moved from `Blockly` to `Blockly.common` (and other places) was done to resolve circular `requires` and (more recently) so that no `goog.module` need depend directly on the `Blockly` module, rather than because we wanted to change the Blockly API as such—i.e., this code movement was largely an implementation detail.

Of course there will most probably be intentional API changes in the (hopefully near) future as we move towards TypeScript, and I am strongly in favour of trying to make the Blockly API as cleanly organised and modular as possible, but I think it would be well worth nailing down (at least in broad strokes) what we will want that future Blockly API to look like before we start forcing external developers to make arbitrary changes to their code by deprecating everything on `Blockly`.

E.g.: moving the implementation of `defineBlocksWithJsonArray` to `Blockly.common` let us fix problems that would otherwise have occurred when migrating the `blocks/*.js` files, but I didn't understand this to mean that we intended to promulgate a change to the API: there's no obvious reason why external code should not continue to use `Blockly.defineBlocksWithJsonArray` at the moment.

This actually holds true for almost all the other items on `Blockly` deprecated in this and other recent PRs, with the exception of property accessors, where (AFAIK) we do want external code to switch to using the new get/set functions instead—though even here there is no reason in principle that the new functions could not be re-exported from `blockly.js` if we wanted them to be, rather than being called via various sub-modules.

While I am generally in favour of encouraging external developers to update their code as early as possible, I think there is not enough clarity about what the future Blockly API will look like to justify asking developers to make (any but a few) changes at the present time.

maribethb commented:
After discussion, updated which things are actually being deprecated (generally, functions that have a specific and more logical namespace to live):
- bindEvent functions
- clipboard functions
- dialog properties
- some workspace svg related functions that deal with a specific workspace

Did not deprecate (generally, things moved to common.js, don't have an agreed API, or are used in toolbox/block definitions that would be annoying to change):
- alignment, input, connection, and toolbox constants (would affect things like block definitions)
- hideChaff
- createBlocksWithJsonArray
- setParentContainer
- getMainWorkspace and mainWorkspace property
- svgResize

Remaining question:
Blockly.selected is now the only property from blockly.js that is located in the `Blockly.common` namespace that we are now asking developers to use. (It was already marked deprecated not by this PR) Is that what we want to commit to given that as of now everything in Blockly.common is just there to avoid circular deps internally? Or should we wait and see if we have a better namespace for it / allow it to continue to exist on the Blockly namespace?

maribethb commented:
update after confirming with Rachel: removed Blockly.selected from being marked as deprecated. we should deprecate it at some point but going to wait until we have firmed up the location of it (ie not necessarily in common.js)

rachel-fenichel commented:
Approve, and this has addressed Chris's comment.
